internal/pkg/deploy/pipeline.go

```diff
@@ -523,43 +533,74 @@ func (stg *PipelineStage) EnvManagerRoleARN() string {
 
 // Test returns a test for the stage.
 // If the stage does not have any test commands, then returns nil.
-func (stg *PipelineStage) Test() *TestCommandsAction {
+func (stg *PipelineStage) Test() (*TestCommandsAction, error) {
 	if len(stg.testCommands) == 0 {
-		return nil
+		return nil, nil
 	}
 
 	var prevActions []orderedRunner
-	for _, deployment := range stg.Deployments() {
-		prevActions = append(prevActions, &deployment)
+	deployActions, err := stg.Deployments()
+	if err != nil {
+		return nil, err
+	}
+	for _, action := range deployActions {
+		prevActions = append(prevActions, &action)
 	}
 
 	return &TestCommandsAction{
 		action: action{
 			prevActions: prevActions,
 		},
 		commands: stg.testCommands,
-	}
+	}, nil
 }
 
 // Deployments returns a list of deploy actions for the pipeline.
-func (stg *PipelineStage) Deployments() []WorkloadDeployAction {
+func (stg *PipelineStage) Deployments() ([]DeployAction, error) {
 	var prevActions []orderedRunner
 	if approval := stg.Approval(); approval != nil {
 		prevActions = append(prevActions, approval)
 	}
 
-	var actions []WorkloadDeployAction
-	for _, workload := range stg.localWorkloads {
-		actions = append(actions, WorkloadDeployAction{
+	topo, err := graph.TopologicalSort(stg.buildDeploymentsGraph())
+	if err != nil {
+		return nil, fmt.Errorf("topological sort deployments: %v", err)
+	}
+
+	var actions []DeployAction
+	for name, conf := range stg.deployments {
+		actions = append(actions, DeployAction{
 			action: action{
 				prevActions: prevActions,
 			},
-			name:    workload,
-			envName: stg.associatedEnvironment.Name,
-			appName: stg.AppName,
+			name:     name,
+			envName:  stg.associatedEnvironment.Name,
+			appName:  stg.AppName,
+			override: conf,
+			ranker:   topo,
 		})
 	}
-	return actions
+	return actions, nil
+}
+
+func (stg *PipelineStage) buildDeploymentsGraph() *graph.Graph[string] {
+	var names []string
+	for name, _ := range stg.deployments {
+		names = append(names, name)
+	}
```

Comment:
nit: https://pkg.go.dev/golang.org/x/exp/maps#Keys

Reply:
I keep debating if we want to import the `exp` module but I think I'd rather wait until they're promoted to the stdlib, I'll keep it as is for now

Reply:
Sounds good 👍